components/modules/otahttp.c

```diff
@@ -0,0 +1,205 @@
+
+  return lookup.ip_address;
+}
+
+static bool connect_to_http_server(const char* domain, uint16_t port)
+{
+    ESP_LOGI(TAG, "Server: %s Server Port:%d", domain, port);
+
+    ip_addr_t ip_address = dns_lookup(domain);
+
+    if(ip_address.u_addr.ip4.addr == 0) {
+      ESP_LOGE(TAG, "Failed to resolve IP address");
+      return -2;
+    } else {
+      ESP_LOGI(TAG, "DNS found: %i.%i.%i.%i\n", ip4_addr1(&ip_address.u_addr.ip4), ip4_addr2(&ip_address.u_addr.ip4), ip4_addr3(&ip_address.u_addr.ip4), ip4_addr4(&ip_address.u_addr.ip4) );
+    }
+
+    int http_connect_flag = -1;
+    int socket_id = socket(AF_INET, SOCK_STREAM, 0);
+    if (socket_id == -1) {
+        ESP_LOGE(TAG, "Create socket failed!");
+        return -1;
+    }
+
+    struct sockaddr_in sock_info;
+    memset(&sock_info, 0, sizeof(struct sockaddr_in));
+    sock_info.sin_family = AF_INET;
+    sock_info.sin_addr.s_addr = ip_address.u_addr.ip4.addr;
+    sock_info.sin_port = htons(port);
+
+    // connect to http server
+    http_connect_flag = connect(socket_id, (struct sockaddr *)&sock_info, sizeof(sock_info));
+    if (http_connect_flag == -1) {
+        ESP_LOGE(TAG, "Connect to server failed! errno=%d", errno);
+        close(socket_id);
+        return -3;
+    } else {
+        ESP_LOGI(TAG, "Connected to server");
+        return socket_id;
+    }
+}
+
+static int find_body_offset(const char *http_response, int total_len)
+{
+    int i = 0;
+
+    while (http_response[i] != 0 && i + 3 < total_len) {
+      if(http_response[i++] == '\r' && http_response[i++] == '\n' && http_response[i++] == '\r' && http_response[i++] == '\n') {
+        return i;
+      }
+    }
+
+    return -1;
+}
+
+static int otahttp_download( lua_State *L )
+{
+  if(!ota_started) {
+    OTAHTTP_ERROR(L, "Error: OTA not started");
+  }
+
+  int image_length = 0;
+  char http_response[BUFFSIZE + 1];
+  char write_data[BUFFSIZE + 1];
+
+  esp_err_t err;
+  int stack = 0;
+
+  size_t domain_length = 0;
+  const char* domain = luaL_checklstring(L, ++stack, &domain_length);
+  uint16_t port = luaL_checkinteger(L, ++stack);
+  size_t path_length = 0;
+  const char* path = luaL_checklstring(L, ++stack, &path_length);
+
+  /*connect to http server*/
+  int socket_id = connect_to_http_server(domain, port);
+  if (socket_id < 0) {
+    OTAHTTP_ERROR(L, "Connect to http server failed! errno=%d", socket_id);
+  } else {
+    ESP_LOGI(TAG, "Connected to http server");
+  }
+
+  /*send GET request to http server*/
+  char http_request[64] = {0};
+  sprintf(http_request, "GET %s HTTP/1.1\r\nHost: %s:%d \r\n\r\n", path, domain, port);
+
+  int res = send(socket_id, http_request, strlen(http_request), 0);
+  if (res == -1) {
+    OTAHTTP_ERROR(L, "Send GET request to server failed");
+  } else {
+    ESP_LOGI(TAG, "Send GET request to server succeeded");
+  }
+
+  bool found_body = false, flag = true;
+  int packet_counter = 0;
+
+  /* Deal with all receive packets */
+  while (flag) {
```

Comment:
Again, this would be better off as an asynchronous streaming of the HTTP data. The logic is sound, it just shouldn't block all other Lua activity while it is running.

Reply:
Hmm, not sure how this would be done. A FreeRTOS task and callback to LUA once a result is reached?

Reply:
That would be one option (though you'd need to post an event to the Lua task so the actual Lua callback fires from that context - see `task/include/task/task.h`, it's easy). Another option would be to take a look at how the `net` module handles asynchronous network i/o and go down the same route. Both have merit, I'd be happy to approve either approach at this point.